transform_inline.ml

```diff
@@ -0,0 +1,312 @@
+   list of these pairs such that the inlining happens in a bottom up fashion.
+   If a caller-callee pair is already encountered, then it is not included
+   again. This is because once the callee has been inlined in the caller at a
+   given callsite, then the callsite no longer exists. However, multiple
+   callsites with the same target will result in as many caller-callee pairs.
+   Note - The word reduced is used in the general sense of something being
+   simplified. If a function has been analyzed for inlining, its considered
+   reduced. In reality, inlining expands this function. *)
+  let compute_inlining_order (prog : program) : (identifier * identifier) list =
+    let rec inspect_instr ((ord, clr, vis, red, dep) as acc) instr =
+      match[@warning "-4"] instr with
+      | Call (x, (Simple (Constant (Fun_ref f))), es) ->
+        (* If `f` is already visited in this branch, then there is recursion,
+           so don't add this pair to accumulator.*)
+        if (VarSet.mem f vis) then acc
+        (* If `f` is already analyzed for inlining, then add it but don't go
+           inside it.*)
+        else if (VarSet.mem f red)
+        then (List.cons (clr.name, f, dep) ord, clr, vis, red, dep)
+        (* If `f` is neither visited in this branch, nor has it been analyzed
+           for inlining before, then descend into it and analyze it.*)
+        else
+          let clr' = lookup_fun prog f in
+          let dep' = dep + 1 in
+          let (ord'', _, _, red'', _) = add_callees clr' vis red dep' in
+          let ord''' = (clr.name, f, dep)::ord @ ord'' in
+          let red''' = VarSet.union red red'' in
+          (* Note that visited set is only maintained for a branch while
+             reduced set is carried across branches. *)
+          (ord''', clr, vis, red''', dep)
+      (* Any other instruction returns accumulator unchanged. *)
+      | _ -> acc
+    and add_callees clr vis red dep =
+      (* The `clr` is being visited and reduced. So add it to the two sets. *)
+      let vis' = VarSet.add clr.name vis in
+      let red' = VarSet.add clr.name red in
+      Array.fold_left
+        inspect_instr
+        ([], clr, vis', red', dep)
+        (List.hd clr.body).instrs
+    in
+    (* Sort (caller, callee, depth) triplets in decreasing order of depth. *)
+    let comp (_, _, d1) (_, _, d2) = d2 - d1 in
+    let (ord, _, _, _, _) = add_callees main VarSet.empty VarSet.empty 1 in
+    List.map (fun (clr, cle, _ ) -> (clr, cle) ) (List.sort comp ord)
+  in
+
+  (* Given a list of caller - callee pairs, inline the callee inside the caller
+     until the list is exhausted. Order matters, as only this order will result
+     in a maximally inlined program excluding recursive calls. The final program
+     will only have recursive calls left to be inlined. *)
+  let inline_with order prog =
+    List.fold_left
+      (fun p (clr, cle) -> update_fun p (inline_pair clr cle p))
```

Comment:
The idea was to use `clr` and `cle` for just the identifiers and not the actual objects.